### ISSUE TITLE
Inability to filter EventSource categories defined by LoggingEventSource.Keywords

### ISSUE THREAD

**Issue by rymeskar** [aspnet/Logging#899]
LoggingEventSource should be public because one needs to use the 'inside-class' defined Keywords for EventListeners. This pattern was even visible in tests before. The visibility was bent, by marking the test as a friend assembly.

LoggingEventSource should be also sealed as per this [documentation](https://github.com/Microsoft/dotnet-samples/blob/master/Microsoft.Diagnostics.Tracing/EventSource/docs/EventSource.md).  The reasoning behind is to keep the class hierarchy as trivial as possible.

**Comment by davidfowl** [comment]
> LoggingEventSource should be public because one needs to use the 'inside-class' defined Keywords for EventListeners. This pattern was even visible in tests before. 

Can you show an example. This isn't your typical event source, it's an implementation detail.

**Comment by rymeskar** [comment]
Yes, a good example would be if one wants to only allow the emission of JsonMessage.
This would be achieved through calling `EnableEvents(eventSource, EventLevel.LogAlways, LoggingEventSource.Keywords.JsonMessage);` in the overriden OnEventSourceCreated method of a derived EventListener.

What other way is there to filter the 'categories' so that it does not spam your system unnecessarily? I could not find any without the access to the Keywords.

**Comment by rymeskar** [comment]
@davidfowl 
I have thought about the issue a bit more. And you are right, in this case, the LoggingEventSource should be an implementation detail.

I should have rather named the issue: "Inability to filter EventSource categories defined by LoggingEventSource.Keywords"

Then the initially described issue here is just one of the possible solutions to filtering the categories.

The other (and quite frankly cleaner solution without leaking implementaiton details) is in high-level terms:
to provide this 'category' filter on initialization. Then we do not have to leak the information about how the ETW events are emitted. But the fact that the category definitions (Keywords) have to be accessible to the public still holds.

**Comment by davidfowl** [comment]
I fixed it for you, lets expose those keywords.

cc @pakrym

**Comment by rymeskar** [comment]
Do you mean to:

1. Expose only the Keywords, or
2. Also have the possibility to filter the categories right inside the Extensions.Logging.EventSource package? Without needing to leak the info about how the actual logging events are emitted.

**Comment by pakrym** [comment]
I feel like the entire goal of EventSource logger is to be as detailed as possible by default and then collector decides which events to filter.

So I would expose only the keywords.

**Comment by rymeskar** [comment]
I also like to see detailed detailed events.

I just fail to see the added benefit of emitting the same information just under 3 different 'views'. As JsonMessage, Message and FormattedMessage have all the same information, just differently structured.

Plus there is still this comment in the source file:

    /// In addition the level and FilterSpec argument, you can also set EventSource Keywords.  See the Keywords
    /// definition below, but basically you get to decide if you wish to have
    ///
    ///   * Keywords.Message - You get the event with the data in parsed form.
    ///   * Keywords.JsonMessage - you get an event with the data in parse form but as a JSON blob (not broken up by argument ...)
    ///   * Keywords.FormattedMessage - you get an event with the data formatted as a string
    ///
    /// It is expected that you will turn only one of these keywords on at a time, but you can turn them all on (and get
    /// the same data logged three different ways.

Also by collector you mean the actual program that will collect the ETW?? Or the EventListener case mentioned earlier? As far as I understand, Keyword filtering only during the collection of ETW might be a bit less performant.

**Comment by pakrym** [comment]
> Also by collector you mean the actual program that will collect the ETW?? Or the EventListener case mentioned earlier? As far as I understand, Keyword filtering only during the collection of ETW might be a bit less performant.

Both. Why do you think it would be less performant? Keyword filtering is one of ETW features natively supported by EventSource.

> I just fail to see the added benefit of emitting the same information just under 3 different 'views'. As JsonMessage, Message and FormattedMessage have all the same information, just differently structured.

The benefit is that consumer gets to decide what kind of format it wants to get. If you start disabling output formats in the app you are going to lose this flexibility.

**Comment by rymeskar** [comment]
> Both. Why do you think it would be less performant? Keyword filtering is one of ETW features natively supported by EventSource.

I have not done any benchmarking. But my understanding was that there is a way to not emit some event categories altogether [call to _eventSource.IsEnabled inside  EventSourceLogger]. Which must be inherently more performant, less resource draining than only filtering on the collector.

>The benefit is that consumer gets to decide what kind of format it wants to get. If you start disabling output formats in the app you are going to lose this flexibility. 

IC, that this makes sense for a lot of scenarios. 

My background is that our collector application has only the capability to filter based on EventNames (which might still be feasible for this LoggingEventSource)

**Comment by rymeskar** [comment]
> Both. Why do you think it would be less performant? Keyword filtering is one of ETW features natively supported by EventSource.

So I read up on how the keyword filtering is done. And you are right, there is no performance difference whether one enables/disables the keywords in process or out of process.
